File: scripts/fetch-videos.py
#!/usr/bin/env python3
"""
Fetch all videos from Go Meetup Prague YouTube channel using YouTube API v3
and generate a markdown file with video details.
"""

import os
import json
import sys
from datetime import datetime
from pathlib import Path
from urllib.request import urlopen
from urllib.parse import urlencode


def load_env_file(env_path):
    """Load environment variables from .env file."""
    if not env_path.exists():
        return

    with open(env_path) as f:
        for line in f:
            line = line.strip()
            if not line or line.startswith("#"):
                continue
            if "=" in line:
                key, value = line.split("=", 1)
                os.environ[key.strip()] = value.strip()


# Load environment variables from .env file
env_file = Path(__file__).parent.parent / ".env"
load_env_file(env_file)

GOOGLE_API_KEY = os.getenv("GOOGLE_API_KEY")
CHANNEL_ID = os.getenv("CHANNEL_ID")

if not GOOGLE_API_KEY:
    print("Error: GOOGLE_API_KEY not set in .env file")
    sys.exit(1)

if not CHANNEL_ID:
    print("Error: CHANNEL_ID not set in .env file")
    sys.exit(1)


def fetch_videos():
    """Fetch all videos from the YouTube channel."""
    videos = []
    page_token = None
    
    while True:
        params = {
            "key": GOOGLE_API_KEY,
            "channelId": CHANNEL_ID,
            "part": "snippet",
            "type": "video",
            "maxResults": 50,
            "order": "date",
        }
        
        if page_token:
            params["pageToken"] = page_token
        
        url = f"https://www.googleapis.com/youtube/v3/search?{urlencode(params)}"
        
        try:
            with urlopen(url) as response:
                data = json.loads(response.read().decode())
        except Exception as e:
            print(f"Error fetching videos: {e}")
            sys.exit(1)
        
        if "error" in data:
            print(f"YouTube API error: {data['error']}")
            sys.exit(1)
        
        for item in data.get("items", []):
            video_id = item["id"]["videoId"]
            title = item["snippet"]["title"]
            published_at = item["snippet"]["publishedAt"]
            
            # Parse the published date
            published = datetime.fromisoformat(published_at.replace("Z", "+00:00"))
            
            videos.append({
                "id": video_id,
                "title": title,
                "published": published,
                "url": f"https://www.youtube.com/embed/{video_id}",
                "year": published.year,
            })
        
        page_token = data.get("nextPageToken")
        if not page_token:
            break
    
    return videos


def generate_markdown(videos):
    """Generate markdown content from videos."""
    # Sort videos by date (newest first)
    videos.sort(key=lambda v: v["published"], reverse=True)
    
    # Group videos by year
    videos_by_year = {}
    for video in videos:
        year = video["year"]
        if year not in videos_by_year:
            videos_by_year[year] = []
        videos_by_year[year].append(video)
    
    # Generate markdown
    lines = [
        "# Go Meetup Prague Videos",
        "",
        "Auto-generated list of all videos from the Go Meetup Prague YouTube channel.",
        "",
    ]
    
    # Sort years in descending order
    for year in sorted(videos_by_year.keys(), reverse=True):
        lines.append(f"## {year}")
        lines.append("")
        
        for video in videos_by_year[year]:
            lines.append(f"### {video['title']}")
            lines.append(f"- **Video ID**: {video['id']}")
            lines.append(f"- **Published**: {video['published'].strftime('%Y-%m-%d')}")
            lines.append(f"- **URL**: {video['url']}")
            lines.append("")
    
    return "\n".join(lines)


def main():
    """Main function."""
    print("Fetching videos from Go Meetup Prague YouTube channel...")
    videos = fetch_videos()
    print(f"Found {len(videos)} videos")
    
    markdown = generate_markdown(videos)
    
    output_path = Path(__file__).parent.parent / "data" / "videos.md"
    output_path.parent.mkdir(parents=True, exist_ok=True)
    
    with open(output_path, "w") as f:
        f.write(markdown)
    
    print(f"Successfully wrote {len(videos)} videos to {output_path}")


if __name__ == "__main__":
    main()

File: scripts/fetch-videos.go
package main

import (
	"bufio"
	"context"
	"encoding/json"
	"errors"
	"fmt"
	"html"
	"io"
	"log"
	"net/http"
	"net/url"
	"os"
	"sort"
	"strings"
	"time"
)

const youtubeSearchEndpoint = "https://www.googleapis.com/youtube/v3/search"

type Video struct {
	ID        string
	Title     string
	Published time.Time
	URL       string
	Year      int
}

type videoOutput struct {
	ID        string `json:"id"`
	Title     string `json:"title"`
	Published string `json:"published"`
	URL       string `json:"url"`
}

type yearOutput struct {
	Year   int           `json:"year"`
	Videos []videoOutput `json:"videos"`
}

type videosData struct {
	Years []yearOutput `json:"years"`
}

type youtubeSearchResponse struct {
	Items []struct {
		ID struct {
			VideoID string `json:"videoId"`
		} `json:"id"`
		Snippet struct {
			Title       string `json:"title"`
			PublishedAt string `json:"publishedAt"`
		} `json:"snippet"`
	} `json:"items"`
	NextPageToken string `json:"nextPageToken"`
}

func main() {
	loadEnvFile("../.env")

	apiKey := os.Getenv("GOOGLE_API_KEY")
	channelID := os.Getenv("CHANNEL_ID")
	if apiKey == "" || channelID == "" {
		log.Fatal("GOOGLE_API_KEY and CHANNEL_ID must be set in the environment or ../.env")
	}

	client := &http.Client{Timeout: 20 * time.Second}
	videos, err := fetchVideos(context.Background(), client, youtubeSearchEndpoint, apiKey, channelID)
	if err != nil {
		log.Fatalf("Failed to fetch videos: %v", err)
	}

	outputPath := "../data/videos.json"
	if len(os.Args) > 1 {
		outputPath = os.Args[1]
	}
	if err := writeVideos(outputPath, videos); err != nil {
		log.Fatalf("Failed to write videos: %v", err)
	}

	fmt.Printf("Successfully fetched %d videos and wrote to %s\n", len(videos), outputPath)
}

func fetchVideos(ctx context.Context, client *http.Client, endpoint, apiKey, channelID string) ([]Video, error) {
	var videos []Video
	pageToken := ""

	for {
		response, err := fetchPage(ctx, client, endpoint, apiKey, channelID, pageToken)
		if err != nil {
			return nil, err
		}

		for _, item := range response.Items {
			published, err := time.Parse(time.RFC3339, item.Snippet.PublishedAt)
			if err != nil {
				return nil, fmt.Errorf("parse published date %q: %w", item.Snippet.PublishedAt, err)
			}
			if item.ID.VideoID == "" {
				continue
			}
			videos = append(videos, Video{
				ID:        item.ID.VideoID,
				Title:     html.UnescapeString(item.Snippet.Title),
				Published: published,
				URL:       "https://www.youtube.com/embed/" + item.ID.VideoID,
				Year:      published.Year(),
			})
		}

		if response.NextPageToken == "" {
			break
		}
		pageToken = response.NextPageToken
	}

	sort.Slice(videos, func(i, j int) bool {
		return videos[i].Published.After(videos[j].Published)
	})
	return videos, nil
}

func fetchPage(ctx context.Context, client *http.Client, endpoint, apiKey, channelID, pageToken string) (youtubeSearchResponse, error) {
	requestURL, err := url.Parse(endpoint)
	if err != nil {
		return youtubeSearchResponse{}, fmt.Errorf("parse YouTube endpoint: %w", err)
	}
	query := requestURL.Query()
	query.Set("key", apiKey)
	query.Set("channelId", channelID)
	query.Set("part", "snippet")
	query.Set("type", "video")
	query.Set("maxResults", "50")
	query.Set("order", "date")
	if pageToken != "" {
		query.Set("pageToken", pageToken)
	}
	requestURL.RawQuery = query.Encode()

	request, err := http.NewRequestWithContext(ctx, http.MethodGet, requestURL.String(), nil)
	if err != nil {
		return youtubeSearchResponse{}, fmt.Errorf("create YouTube request: %w", err)
	}
	response, err := client.Do(request)
	if err != nil {
		return youtubeSearchResponse{}, fmt.Errorf("request YouTube API: %w", err)
	}
	defer response.Body.Close()

	if response.StatusCode != http.StatusOK {
		body, readErr := io.ReadAll(io.LimitReader(response.Body, 4096))
		if readErr != nil {
			return youtubeSearchResponse{}, fmt.Errorf("YouTube API returned %s", response.Status)
		}
		return youtubeSearchResponse{}, fmt.Errorf("YouTube API returned %s: %s", response.Status, strings.TrimSpace(string(body)))
	}

	var result youtubeSearchResponse
	if err := json.NewDecoder(response.Body).Decode(&result); err != nil {
		return youtubeSearchResponse{}, fmt.Errorf("decode YouTube response: %w", err)
	}
	return result, nil
}

func writeVideos(path string, videos []Video) error {
	years := make(map[int][]videoOutput)
	for _, video := range videos {
		years[video.Year] = append(years[video.Year], videoOutput{
			ID:        video.ID,
			Title:     video.Title,
			Published: video.Published.Format("2006-01-02"),
			URL:       video.URL,
		})
	}

	yearNumbers := make([]int, 0, len(years))
	for year := range years {
		yearNumbers = append(yearNumbers, year)
	}
	sort.Sort(sort.Reverse(sort.IntSlice(yearNumbers)))

	data := videosData{Years: make([]yearOutput, 0, len(yearNumbers))}
	for _, year := range yearNumbers {
		data.Years = append(data.Years, yearOutput{Year: year, Videos: years[year]})
	}

	encoded, err := json.MarshalIndent(data, "", "  ")
	if err != nil {
		return fmt.Errorf("encode videos: %w", err)
	}
	encoded = append(encoded, '\n')
	if err := os.WriteFile(path, encoded, 0o644); err != nil {
		return fmt.Errorf("write %s: %w", path, err)
	}
	return nil
}

func loadEnvFile(path string) {
	file, err := os.Open(path)
	if err != nil {
		if !errors.Is(err, os.ErrNotExist) {
			log.Printf("Warning: could not open %s: %v", path, err)
		}
		return
	}
	defer file.Close()

	scanner := bufio.NewScanner(file)
	for scanner.Scan() {
		line := strings.TrimSpace(scanner.Text())
		if line == "" || strings.HasPrefix(line, "#") {
			continue
		}
		parts := strings.SplitN(line, "=", 2)
		if len(parts) != 2 {
			continue
		}
		key := strings.TrimSpace(parts[0])
		value := strings.TrimSpace(parts[1])
		if _, exists := os.LookupEnv(key); !exists && value != "" {
			if err := os.Setenv(key, value); err != nil {
				log.Printf("Warning: could not set %s: %v", key, err)
			}
		}
	}
	if err := scanner.Err(); err != nil {
		log.Printf("Warning: could not read %s: %v", path, err)
	}
}
